core/src/main/java/com/datastax/oss/driver/api/core/type/codec/registry/CodecRegistry.java

```diff
@@ -70,6 +70,26 @@
    */
   <T> TypeCodec<T> codecFor(DataType cqlType);
 
+  /**
+   * Returns a codec to convert the given Java type to the CQL type deemed most appropriate to
+   * represent it.
+   *
+   * <p>The definition of "most appropriate" is unspecified, and left to the appreciation of the
+   * registry implementor.
+   */
+  <T> TypeCodec<T> codecFor(GenericType<T> javaType);
```

Comment:
I still have doubts about this, in particular with the fact that the implementation _happens to be_ covariant. Is this an implementation detail, or should we make it part of the contract? Or should it actually be invariant, and require a dedicated implementation instead of reusing what we had previously?

It doesn't help that we have no identified use case for this method. You mentioned people in the community requested it, do you remember when that conversation happened?

I just don't want this to become a "it sounded like a good idea at the time" moment 😉

Reply:
The discussion has been captured in [JAVA-1015](https://datastax-oss.atlassian.net/browse/JAVA-1015).

The original pull request had an invariant implementation (i.e. it used `codec.accepts(javaType)`, not `codec.getJavaType().isSuperTypeOf(javaType)`).

Indeed I'd rather keep it invariant. Suppose that it's covariant, that we have `B extends A` and there is a codec for `A`. The following would happen:

```
BoundStatement s = ...
s.set(0, new B(), codecFor(B.class)); // ok, will return TypeCodec<A> which can encode B instances
s.get(0, codecFor(B.class)); // TypeCodec<A> will decode B to an A instance -> ClassCastException
```

But I don't think we can actually. 

Because there is something new in 4.x: previously lookups by value wouldn't be cached, but now they are. 

What was your motivation to change that?

The practical consequence is that lookups by java type now _have_ to be covariant, whether or not the initial entry point was a lookup by value or by java type:

1. `codecFor(Object)` or `codecFor(GenericType<?>)` may both call
2. `getCachedCodec(DataType, GenericType<?>)` which may call
3. `createCodec(DataType, GenericType<?>)` which may call (since `cqlType` is null)
4. `createCovariantCodec(GenericType<?> javaType)`, which may call
5. `codecFor(GenericType<T> javaType)` (previously named `covariantCodecFor(GenericType<T> javaType)`)

... which must be covariant because the entry point above might have been `codecFor(Object)`, which requires covariance.

I only see two ways to fix this issue:

1. Do not cache lookups by value, use covariant lookups only for these lookups.
2. Keep caching lookups by value, but store them in a separate cache that creates covariant codecs.

I will provide a solution using approach 1 above, as `codecFor(Object)` is only used in `SimpleStatement`s and I don't think that performance is paramount.